Don't use generic serialization automatically, add a test

fingolfin opened the pull request:
For some types (e.g. from Nemo, Singular) it can produce nonsense,
even corrupt objects and ultimately crashes.

Instead, allow types to opt-in to this generic handling. This
could be made even nicer with a macro if so desired. So that one could write, say
```julia
@serialize_generic PolyhedralComplex
```
instead of
```
save_internal(s::SerializerState, obj::PolyhedralComplex) = save_internal_generic(s, obj)
load_internal(s::DeserializerState, ::Type{T}, dict::Dict) where T <: PolyhedralComplex = load_internal_generic(s, T, dict)
```
I'd be happy to write such a macro, but before doing that, I'd rather first find out what others think of this PR :-)

fieker commented on src/Serialization/main.jl:
```diff
@@ -166,7 +166,7 @@ end
+function save_internal_generic(s::SerializerState, obj::T) where T
     result = Dict{Symbol, Any}()
     for n in fieldnames(T)
         if n != :__attrs
```
Why don't you want to serialise the attributes?

fingolfin replied:
Good question. But I deliberately didn't want to touch it with this PR, as it is orthogonal to the concern this PR is supposed to address :-)

lkastner replied:
I skipped them as they didn't work (for toric varieties). I'd like to have them in the future.

fieker commented on test/Serialization/PolyhedralGeometry.jl:
```diff
@@ -32,7 +32,20 @@
+            save(PC, joinpath(path, "docu.pc"))
+            loaded = load(joinpath(path, "docu.pc"))
+            @test loaded isa PolyhedralComplex
+            @test Base.propertynames(PC) == Base.propertynames(loaded)
```
is there any way to create some idempotent test? e.g O object. L = load(save(O)) and verify save(L) == save(O)? Is this desirable? This would idaly require to save into memory rather than a file and would go hand-in-hand with a different interface: instead of a file name use some io-context...

fingolfin replied:
In principle one could of course check `PC == loaded` here -- assuming `==` is implemented for these types, which I don't know and didn't check. Rather, I simply copy&pasted and slightly adapted one of the existing tests :-).

As to saving to and loading from io context, see issue #1319

fingolfin replied:
Aha, I misunderstood, you want to save/load/save again, and compare the two saved objects. Sounds good! One problem are the UUIDs which will differ (see also issue #1340)

lkastner replied:
Didn't you point out to me that Julia dicts aren't ordered? So loading we might get a different order and then saving will produce something different independent of the UUID issue.

fieker reviewed (commented):
apart from the 2 general comments - to whomever, i regard tis more as a discussion than anything else.
Using the generic code by default is extremely error prone and should not be done. A better way might be to use some meta-programming to do s.th. like
  for T in list_of_types
    save(a::T) = generic_save(a)
    load(::T) = generic_load()

fingolfin commented:
> apart from the 2 general comments - to whomever, i regard tis more as a discussion than anything else.
> Using the generic code by default is extremely error prone and should not be done. A better way might be to use some meta-programming to do s.th. like
> for T in list_of_types
> save(a::T) = generic_save(a)
> load(::T) = generic_load()

But that's exactly what this PR is about?

fieker commented:
On Tue, May 17, 2022 at 06:41:04AM -0700, Max Horn wrote:
> > apart from the 2 general comments - to whomever, i regard tis more as a discussion than anything else.
> > Using the generic code by default is extremely error prone and should not be done. A better way might be to use some meta-programming to do s.th. like
> > for T in list_of_types
> > save(a::T) = generic_save(a)
> > load(::T) = generic_load()
> 
> But that's exactly what this PR is about? 
> 
You listed the save and load functions individualy - I was suggesting to
literally do 
  for T in [types that are automatic]
    eval(Meta....)
 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/oscar-system/Oscar.jl/pull/1341#issuecomment-1128885348
> You are receiving this because your review was requested.
> 
> Message ID: ***@***.***>

fingolfin commented:
> You listed the save and load functions individualy - I was suggesting to literally do for T in [types that are automatic] eval(Meta....)

I did this in an earlier version of this PR by deliberately changed it, as I felt it did not make things clearer, and obscured the extensibility... Indeed, on the long run, I am not sure it makes sense to concentrate all serialization functions inside `src/Serialize`; instead it might make more sense to put them where other methods for the respective types are (of course with the exception of "basic" types like `fmpz`, which of course need to stay in `src/Serialize/` as their implementation is in another package).

Anyway, that's trivial to change either way

lkastner commented:
Even though I would lean more in the direction of an opt-out approach, I am ok with an opt-in macro. Thank you.

fingolfin committed:
Don't use generic serialization automatically

For some types (e.g. from Nemo, Singular) it can produce nonsense,
even corrupt objects and ultimately crashes.

Instead, allow types to opt-in to this generic handling. This
could be made even nicer with a macro if so desired.

fieker commented:
merge? Anyone objecting?
